crux-core/src/crux/kv_indexer.clj

```diff
@@ -109,6 +109,106 @@
+   (assert (or (zero? (.capacity content-hash)) (= c/id-size (.capacity content-hash)))
+           (mem/buffer->hex content-hash))
+   (let [^MutableDirectBuffer b (or b (mem/allocate-buffer (+ c/index-id-size (.capacity attr) (.capacity entity) (.capacity content-hash) (.capacity v))))]
+     (-> (doto b
+           (.putByte 0 c/aecv-index-id)
+           (.putBytes c/index-id-size attr 0 (.capacity attr))
+           (.putBytes (+ c/index-id-size (.capacity attr)) entity 0 (.capacity entity))
+           (.putBytes (+ c/index-id-size (.capacity attr) (.capacity entity)) content-hash 0 (.capacity content-hash))
+           (.putBytes (+ c/index-id-size (.capacity attr) (.capacity entity) (.capacity content-hash)) v 0 (.capacity v)))
+         (mem/limit-buffer (+ c/index-id-size (.capacity attr) (.capacity entity) (.capacity content-hash) (.capacity v)))))))
+
+(defn decode-aecv-key-from ^crux.kv_indexer.Quad [^DirectBuffer k]
+  (let [length (long (.capacity k))]
+    (assert (<= (+ c/index-id-size c/id-size c/id-size) length) (mem/buffer->hex k))
+    (let [index-id (.getByte k 0)]
+      (assert (= c/aecv-index-id index-id))
+      (let [value-size (- length c/id-size c/id-size c/id-size c/index-id-size)
+            attr (Id. (mem/slice-buffer k c/index-id-size c/id-size) 0)
+            entity (Id. (mem/slice-buffer k (+ c/index-id-size c/id-size) c/id-size) 0)
+            content-hash (Id. (mem/slice-buffer k (+ c/index-id-size c/id-size c/id-size) c/id-size) 0)
+            value (mem/slice-buffer k (+ c/index-id-size c/id-size c/id-size c/id-size) value-size)]
+        (->Quad attr entity content-hash value)))))
+
+(defn all-attrs [i]
+  (let [seek-buffer (.get idx/seek-buffer-tl)
+        aecv-prefix (encode-aecv-key-to seek-buffer)
+        i (new-prefix-kv-iterator i aecv-prefix)]
+    (letfn [(step [k]
+              (lazy-seq
+               (when-let [^DirectBuffer k (kv/seek i k)]
+                 (cons (Id. (mem/slice-buffer k c/index-id-size c/id-size) 0)
+                       (step (-> (mem/copy-buffer k (.capacity k) seek-buffer)
+                                 (mem/limit-buffer (+ c/index-id-size c/id-size))
+                                 (mem/inc-unsigned-buffer!)))))))]
+      (step aecv-prefix))))
+
+(defn encode-hash-cache-key-to
+  (^org.agrona.MutableDirectBuffer [b entity]
+   (encode-aecv-key-to b entity c/empty-buffer))
+
+  (^org.agrona.MutableDirectBuffer [^MutableDirectBuffer b ^DirectBuffer entity ^DirectBuffer value]
+   (assert (= c/id-size (.capacity entity)) (mem/buffer->hex entity))
+
+   (let [^MutableDirectBuffer b (or b (mem/allocate-buffer (+ c/index-id-size c/id-size (.capacity value))))]
+     (-> (doto b
+           (.putByte 0 c/hash-cache-index-id)
+           (.putBytes c/index-id-size entity 0 c/id-size)
+           (.putBytes (+ c/index-id-size c/id-size) value 0 (.capacity value)))
+         (mem/limit-buffer (+ c/index-id-size c/id-size (.capacity value)))))))
+
+;;;; Entity as-of
+
```

Comment:
At what point do we make a new `kv_index_store.clj` namespace?

Reply:
(discussed offline) this namespace is dealing with both the write and the read to the kv-store-backed indexer/index-store - the Indexer largely being the write side; the IndexStore being the read. 

Arguably it could be renamed to `crux.kv-index-store` but at that point I'd probably rename the `Indexer` protocol to `IndexStore`, and `IndexStore` to ... something else.